### PR TITLE
#170947605: Users should be able to Login

### PR DESCRIPTION
**Pull request by Kagorora**
#### What does this PR do?

- Allow users to Login in the system

#### Description of Task to be completed?

- [x] Added Login fronted design

- [x] Added Login Redux functionality

- [x] Added Error Handling

- [x] Added Input validations

- [x] Test components

#### How should this be manually tested?

- clone the project by running `git clone https://github.com/Stackup-Rwanda/knights-bn-frontend.git`

- open the project `cd knights-bn-frontend` then `code .`

- Install dependencies `npm install`

- Run the project `npm run dev-start`

- click on login on the nav-bar

#### Heroku Link

       https://knights-bn-f-ft-login-1-aujgoc.herokuapp.com/

#### Mockup Link

       https://xd.adobe.com/view/ddb6dde0-8b48-4220-6a0f-493993ea9646-ab87/

#### Any background context you want to provide?

- N/A

#### What are the relevant pivotal tracker stories?

- [170947605](https://www.pivotaltracker.com/story/show/170947605)

#### Screenshots (if appropriate)

Screen Shot 2020-04-30 at 09 42 14

#### Questions:

- N/A

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -21,3 +21,9 @@ jobs:
     - run:
         name: coveralls
         command: npm run coveralls
+workflows:
+  version: 2
+  workflow:
+    jobs:
+    - build
+
```
Comment (kagabof) [inline .circleci/config.yml]: Add a newline in this file.

File: src/components/Alert.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { Alert } from 'react-bootstrap';
+
+const AlertComponent = (props) => {
+  const { variant, alertMsg } = props;
```
Comment (kagabof) [inline src/components/Alert.js]: You can destructure directly in the function parameters without creating a new variable.

File: src/components/Button.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Button } from 'react-bootstrap';
+
+function SubmitButton(props) {
```
Comment (kagabof) [inline src/components/Button.js]: please use S6 function syntax.

File: src/components/Nav.js [hunk 4]
```diff
@@ -7,14 +7,16 @@ const Nav = () => {
   return (
```
Comment (kagabof) [inline src/components/Nav.js]: There is no need of return statement.

File: src/redux/actions/actions.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import axios from 'axios';
+import {
+  FETCH_USERS_REQUEST,
+  FETCH_USERS_SUCCESS,
+  FETCH_USERS_FAILURE,
+  VALIDATE_INPUT,
+  USER_LOGIN_SUCCESS,
+} from './actionTypes';
```
Comment (kagabof) [inline src/redux/actions/actions.js]: You can use a stare to import all functions in that file. ex: ```import * as type from './actionTypes' ```

File: src/redux/reducers/userReducer.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import {
+  FETCH_USERS_REQUEST,
+  FETCH_USERS_SUCCESS,
+  FETCH_USERS_FAILURE,
+  USER_LOGIN_SUCCESS,
+} from '../actions/actionTypes';
```
Comment (kagabof) [inline src/redux/reducers/userReducer.js]: You can use a stare to import all functions in that file. ex: ```import * as type from './actionTypes'```

File: src/views/Home.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+export default function Home() {
```
Comment (kagabof) [inline src/views/Home.js]: please use es6 syntax.

**Review by T2Wil — DISMISSED** [review]

File: src/views/Login.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+class Login extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      data: { email: "", password: "" },
+      error: " ",
+      loggedIn: false,
+    };
+  }
```
Comment (kagabof) [inline src/views/Login.js]: In es6 you don't need a constructor.

**Review by Niyonsengaeric — DISMISSED** [review]

File: src/views/Login.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+          <p>{this.props.error}</p>
+          <Button onClick={(e) => this.login(e)} id="loginBtn" label="Login" />
+          <a id="forgotPassword" href="#">
+            <p>Forgot Password?</p>
+          </a>
+        </Form>
+      </div>
+    );
+  }
+  login = async (e) => {
```
Comment (kagabof) [inline src/views/Login.js]: Just put this function before render method.

File: src/views/Login.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      <div className="login-page">
+        <Form>
+          <h1>Sign in</h1>
+          <span>
+            <p className="text-danger" id="loginError">
+              {error ? error : " "}
+            </p>
+          </span>
+          <p>Login with so social media</p>
+          <div className="Google-login-btn">
```
Comment (kagabof) [inline src/views/Login.js:75]: Please refactor this file by making some of the parts of its components.

File: src/views/LoginPage.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import Login from './Login';
+
+export default function LoginPage() {
```
Comment (kagabof) [inline src/views/LoginPage.js]: Please use the es6 syntax.

**Review by kagabof — CHANGES_REQUESTED** [review]
Please make sure you solved those issues.

**Review by EugeneMunya — DISMISSED** [review]

File: src/views/Login.js (hunk 10 above)
Comment (Kagorora) [reply src/views/Login.js:75]: @kagabof This will be done when implementing the social login

File: __tests__/Footer.test.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { cleanup } from '@testing-library/react';
+import '@testing-library/jest-dom';
+import Footer from '../src/components/Footer';
+
+describe('=============> Footer <=============', () => {
```
Comment (kagabof) [inline __tests__/Footer.test.js]: Please describe properly.

File: __tests__/Header.test.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { render, cleanup } from '@testing-library/react';
+import '@testing-library/jest-dom';
+import Header from '../src/components/Header';
+
+describe('===== Test Header =====', () => {
```
Comment (kagabof) [inline __tests__/Header.test.js]: Please describe properly.

File: __tests__/Login.test.js [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+  };
+
+
+  it('Should not pass, empty inputs', () => {
+    wrapper = mount(
+      <MemoryRouter>
+        <View {...props} />
+      </MemoryRouter>,
+    );
+    console.log('<<<<<</////>>>>', wrapper.props());
```
Comment (kagabof) [inline __tests__/Login.test.js]: remove those consoles

File: __tests__/Login.test.js [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    console.log('<<<<<', wrapper.find('Button').simulate('click'));
```
Comment (kagabof) [inline __tests__/Login.test.js]: remove this console.

File: __tests__/Login.test.js [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  };
+
+  it('Should not pass, empty password', () => {
+    wrapper = mount(
+      <MemoryRouter>
+        <View {...props} />
+      </MemoryRouter>,
+    );
+    console.log('<<<<<</////>>>>', wrapper.props());
+    console.log('<<<<<', wrapper.find('Button').simulate('click'));
```
Comment (kagabof) [inline __tests__/Login.test.js]: remove those console

File: __tests__/__snapshots__/Nav.spec.js.snap [hunk 17]
```diff
@@ -0,0 +1,44 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (kagabof) [inline __tests__/__snapshots__/Nav.spec.js.snap]: please ignore those snapshots.

**Review by kagabof — CHANGES_REQUESTED** [review]

File: Untitled-1 [hunk 18]
```diff
@@ -0,0 +1,17 @@
+module.export = {
```
Comment (kagabof) [inline Untitled-1]: Name property this file.

File: src/jest.config.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
+// module.exports = {
```
Comment (kagabof) [inline src/jest.config.js]: why do we have commented code?

**Review by kagabof — CHANGES_REQUESTED** [review]

**Review by kagabof — CHANGES_REQUESTED** [review]
Why do we have two jest.config.js?
I need you to test you to increase your test coverage, it has to be above 99%.

**Review by kagabof — APPROVED** [review]
Well done.

**Pull request merged**